src/ui/wxWidgets/DragBarCtrl.cpp

```diff
@@ -96,78 +95,211 @@ struct _DragbarElementInfo {
-    }
+  ClearTools();
+
+  for (int idx = 0; size_t(idx) < NumberOf(DragbarElements); ++idx) {
+    AddTool(
+      idx + DRAGBAR_TOOLID_BASE,
+      BUTTON_ENABLED, BUTTON_DISABLED,
+      false, nullptr,
+      BUTTON_TOOLTIP
+    );
   }
-  else {
+
+  Realize();
+
+#undef BUTTON_ENABLED
+#undef BUTTON_DISABLED
+#undef BUTTON_TOOLTIP
+}
+
+void DragBarCtrl::UpdateBitmaps()
+{
+  const bool newButtons = PWSprefs::GetInstance()->GetPref(PWSprefs::UseNewToolbar);
+
+#define BUTTON_ENABLED newButtons ? \
+  wxBitmap(DragbarElements[idx].bitmap) : \
+  wxBitmap(DragbarElements[idx].classic_bitmap)
+#define BUTTON_DISABLED newButtons ? \
+  wxBitmap(DragbarElements[idx].bitmap_disabled) : \
+  wxBitmap(DragbarElements[idx].classic_bitmap_disabled)
+
+  if (GetToolCount() > 0) {
     for (int idx = 0; size_t(idx) < NumberOf(DragbarElements); ++idx) {
-      SetToolBitmaps(idx + DRAGBAR_TOOLID_BASE, BTN, BTN_DISABLED);
+      auto tool = FindToolByIndex(idx);
+      if (tool) {
+        tool->SetBitmap(BUTTON_ENABLED);
+        tool->SetDisabledBitmap(BUTTON_DISABLED);
+      }
     }
+
+    Realize();
   }
 
-#undef BTN
-#undef BTN_DISABLED
+#undef BUTTON_ENABLED
+#undef BUTTON_DISABLED
 }
 
-DragBarCtrl::~DragBarCtrl()
+void DragBarCtrl::UpdateTooltips()
 {
+#define BUTTON_TOOLTIP ((idx == DND_IDX) ? \
+  wxString(_("Drag this image onto another window to paste the selected element or tree.")) : \
+  wxString(_("Drag this image onto another window to paste the '")) << _(DragbarElements[idx].name) << _("' field."))
+
+  if (GetToolCount() > 0) {
+    for (int idx = 0; size_t(idx) < NumberOf(DragbarElements); ++idx) {
+      auto tool = FindToolByIndex(idx);
+      if (tool) {
+        tool->SetShortHelp(BUTTON_TOOLTIP);
+      }
+    }
+
+    Realize();
+  }
+
+#undef BUTTON_TOOLTIP
 }
 
-wxString DragBarCtrl::GetText(int id) const
+wxString DragBarCtrl::GetText(int idx) const
 {
-  const int idx = id - DRAGBAR_TOOLID_BASE;
   wxASSERT( idx >= 0 && size_t(idx) < NumberOf(DragbarElements));
 
-  if(idx == DND_IDX)
+  if(idx == DND_IDX) {
     return wxString(wxEmptyString);
-  
+  }
+
   const CItemData *pci(nullptr), *pbci(nullptr);
-  pci = m_frame->GetSelectedEntry();
-  pbci = m_frame->GetBaseEntry(pci);
+  auto mainFrame = wxGetApp().GetPasswordSafeFrame();
+  wxASSERT(mainFrame);
+  pci = mainFrame->GetSelectedEntry();
+  pbci = mainFrame->GetBaseEntry(pci);
 
   return pci ?
     towxstring(pci->GetEffectiveFieldValue(DragbarElements[idx].ft, pbci)) : wxString(wxEmptyString);
 }
 
-bool DragBarCtrl::IsEnabled(int id) const
+void DragBarCtrl::OnDrag(wxAuiToolBarEvent& event)
 {
-  const int idx = id - DRAGBAR_TOOLID_BASE;
-  wxASSERT( idx >= 0 && size_t(idx) < NumberOf(DragbarElements));
-  
-  if(idx == DND_IDX) {
+  auto toolId = static_cast<size_t>(event.GetToolId()) - DRAGBAR_TOOLID_BASE;
+  wxASSERT(toolId >= 0 && toolId <= NumberOf(DragbarElements));
+
+  // Separators have the index -1 (ID_SEPARATOR)
+  if (toolId < 0) {
```

Comment:
In other places we directly compare with `ID_SEPARATOR ` - wouldn't that be more clear?

Reply:
Resolved in fc4ed9d. The check was removed because no separators are used at dragbar.